*Ralstonia eutropha H16* RBA model
================
Michael Jahn,
2020-04-28

***

This resource balance analysis (RBA) model was generated using the [RBApy package from Bulovic et al.](https://doi.org/10.1016/j.ymben.2019.06.001),
using a manually curated metabolic reconstruction originally developed by [Park et al., 2011](http://bmcsystbiol.biomedcentral.com/articles/10.1186/1752-0509-5-101). The original model was refactored and many reactions were curated and harmonized to community standards as e.g. available through the [Bigg model database](bigg.ucsd.edu/). The genome scale model that serves as base for the RBA model is maintained on [github.com/m-jahn/](https://github.com/m-jahn/genome-scale-models).

Sources for parameter estimation are:

 - [Goelzer at al., 2015](https://doi.org/10.1016/j.ymben.2015.10.003): set of parameters for transcription, translation, and other fundamental cellular functions.


## Generation of the XML model

The *Ralstonia eutropha H16* RBA model was generated in a step-wise manner,
based on information from a hand-curated model.


### Step 1: generation of a default model

For initial creation of the RBA model, the following input files were used. The fasta file headers have to be formatted according to this pattern: `rba|id|name|set_name|stoichiometry`.

 - `data/sbml.xml`: metabolic netwok from the hand-curated model.
 - `data/ribosome.fasta`: composition of ribosome (rRNA and proteins). rRNA sequences were retrieved from NCBI, ribosomal protein sequence and subunit structure from uniprot. 68 ribosomal proteins were included. Ribosomal stoichiometry was assumed to be one subunit each for 30S and 50S subunits, except for L7/L12 that forms 2-4 dimers per ribosome (`n >= 4`). *TODO: update stochiometry for elongation and initiation factors.*
 - `data/chaperones.fasta`: composition of molecular machines involved in protein folding. These include the protein subunits `groES`, `groEL`, `dnaJ`, `dnaK`, `grpE`, `trigger factor tig`, `htpG`, and `hfq` with their respective number od subunits obtained from uniprot.
 - `data/trnas.fasta`: composition of tRNAs. tRNA sequences were retrieved from NCBI, and only one representative tRNA (the first) per amino acid was used (20 out of 61 tRNAs).
 - `params.in`: pipeline parameters.

Then we start the initial RBApy model generation via the following customized function. The function automatically applies custom changes to the SBML model, imports the `params.in` file with basic settings, and generates the model files.

```
python generate_model.py
```

This first run downloaded `data/uniprot.csv`, generated helper files `data/*.tsv`, and generated a first XML model. This model can be solved by running

```
python solve_rba_model.py .
```

### Step 2: generation of a model improved through helper files

In order to improve RBA's predictions, the following input files were manually curated:

 - `macrocomponents.tsv`: added concentration targets for `PHOSPHOLIPIDS_c`, `LPS_c` (lipopolysaccharide), `PETIDOGLYCAN_c`, taken from model's biomass equation.
 - `medium.tsv`: added new minimal medium.
 - `metabolites.tsv`: mapped the metabolites used in processes and added concentration targets for metabolites (TODO: update target concentrations).
 - `subunits.tsv`: updated some protein stoichiometries (ATP synthase).
 - `unknown_proteins.tsv`: (empty) the SBML model contains no genes that could not be mapped

### Step 3: generation of the full model


In the last step, the full model can be generated by running one more iteration of:
```
  python generate_model.py
```


## Differences between the hand-curated model and the model generated by RBApy

To be added.
